File: Android/app/src/main/java/com/arcinteractive/spaces/data/spaces/LinkPreviewService.kt
package com.arcinteractive.spaces.data.spaces

import android.content.Context
import android.graphics.BitmapFactory
import android.util.Base64
import android.util.Patterns
import com.arcinteractive.spaces.data.model.LinkPreviewData
import com.arcinteractive.spaces.data.model.SpaceLinkAttachment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL

data class EncryptedTextMessageContent(
    val version: Int = 1,
    val text: String,
    val linkPreview: LinkPreviewData? = null,
    val spaceLinks: List<SpaceLinkAttachment>? = null
)

class LinkPreviewService {
    private val cacheTtlMillis = 7L * 24L * 60L * 60L * 1000L

    fun firstUrl(text: String): String? {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return null

        val matcher = Patterns.WEB_URL.matcher(trimmed)
        while (matcher.find()) {
            val raw = matcher.group().orEmpty()
            val normalized = normalizeUrl(raw) ?: continue
            return normalized
        }

        trimmed.split(Regex("\\s+")).forEach { token ->
            normalizeUrl(token)?.let { return it }
        }
        return null
    }

    suspend fun cachedPreview(context: Context, url: String): LinkPreviewData? = withContext(Dispatchers.IO) {
        val key = cacheKey(url)
        val json = cachePrefs(context).getString(key, null) ?: return@withContext null
        val parsed = runCatching { JSONObject(json) }.getOrNull() ?: return@withContext null
        val cachedAt = parsed.optLong("cachedAt", 0L)
        if (cachedAt <= 0L || System.currentTimeMillis() - cachedAt > cacheTtlMillis) {
            cachePrefs(context).edit().remove(key).apply()
            return@withContext null
        }
        parsePreviewJson(parsed.optJSONObject("preview"))
    }

    suspend fun preview(context: Context, url: String): LinkPreviewData? = withContext(Dispatchers.IO) {
        cachedPreview(context, url)?.let { return@withContext it }

        val normalizedUrl = normalizeUrl(url) ?: return@withContext null
        val htmlResult = fetchText(normalizedUrl) ?: return@withContext null
        val metadata = parseMetadata(htmlResult.html, normalizedUrl)
        val title = sanitize(metadata.title, 120) ?: return@withContext null
        val previewImage = metadata.imageUrl?.let { fetchImageBytes(it) }
        val preview = LinkPreviewData(
            originalUrl = normalizedUrl,
            canonicalUrl = metadata.canonicalUrl,
            domain = domain(metadata.canonicalUrl ?: normalizedUrl),
            title = title,
            summary = sanitize(metadata.description, 220),
            siteName = sanitize(metadata.siteName, 80),
            imageDataBase64 = previewImage?.first?.let { Base64.encodeToString(it, Base64.NO_WRAP) },
            imageMimeType = previewImage?.second
        )
        storePreview(context, normalizedUrl, preview)
        preview
    }

    private fun cachePrefs(context: Context) =
        context.applicationContext.getSharedPreferences("spaces_link_preview_cache", Context.MODE_PRIVATE)

    private fun storePreview(context: Context, url: String, preview: LinkPreviewData) {
        val json = JSONObject().apply {
            put("cachedAt", System.currentTimeMillis())
            put("preview", JSONObject().apply {
                put("originalUrl", preview.originalUrl)
                put("canonicalUrl", preview.canonicalUrl)
                put("domain", preview.domain)
                put("title", preview.title)
                put("summary", preview.summary)
                put("siteName", preview.siteName)
                put("imageDataBase64", preview.imageDataBase64)
                put("imageMimeType", preview.imageMimeType)
            })
        }
        cachePrefs(context).edit().putString(cacheKey(url), json.toString()).apply()
    }

    private fun parsePreviewJson(json: JSONObject?): LinkPreviewData? {
        json ?: return null
        val originalUrl = json.optString("originalUrl").ifBlank { return null }
        val domain = json.optString("domain").ifBlank { return null }
        val title = json.optString("title").ifBlank { return null }
        return LinkPreviewData(
            originalUrl = originalUrl,
            canonicalUrl = json.optString("canonicalUrl").ifBlank { null },
            domain = domain,
            title = title,
            summary = json.optString("summary").ifBlank { null },
            siteName = json.optString("siteName").ifBlank { null },
            imageDataBase64 = json.optString("imageDataBase64").ifBlank { null },
            imageMimeType = json.optString("imageMimeType").ifBlank { null }
        )
    }

    private fun cacheKey(url: String): String = normalizeUrl(url)?.lowercase().orEmpty()

    private fun normalizeUrl(raw: String): String? {
        val trimmed = raw.trim().trimEnd('.', ',', ';', ':', '!', '?', ')', ']')
        if (trimmed.isEmpty()) return null
        val withScheme = if (trimmed.startsWith("www.", ignoreCase = true)) {
            "https://$trimmed"
        } else {
            trimmed
        }
        return runCatching {
            val uri = URI(withScheme)
            val scheme = uri.scheme?.lowercase()
            if (scheme == "http" || scheme == "https") uri.toString() else null
        }.getOrNull()
    }

    private data class HtmlFetchResult(
        val finalUrl: String,
        val html: String
    )

    private fun fetchText(urlString: String): HtmlFetchResult? {
        val connection = (URL(urlString).openConnection() as? HttpURLConnection) ?: return null
        return try {
            connection.instanceFollowRedirects = true
            connection.connectTimeout = 15_000
            connection.readTimeout = 15_000
            connection.setRequestProperty("User-Agent", "Spaces/1.1")
            connection.connect()
            if (connection.responseCode !in 200..399) return null
            val finalUrl = connection.url.toString()
            val body = connection.inputStream.buffered().reader(Charsets.UTF_8).use { reader ->
                val buffer = CharArray(256_000)
                val read = reader.read(buffer)
                if (read <= 0) "" else String(buffer, 0, read)
            }
            if (body.isBlank()) null else HtmlFetchResult(finalUrl, body)
        } catch (_: Throwable) {
            null
        } finally {
            connection.disconnect()
        }
    }

    private fun fetchImageBytes(urlString: String): Pair<ByteArray, String>? {
        val connection = (URL(urlString).openConnection() as? HttpURLConnection) ?: return null
        return try {
            connection.instanceFollowRedirects = true
            connection.connectTimeout = 15_000
            connection.readTimeout = 15_000
            connection.setRequestProperty("User-Agent", "Spaces/1.1")
            connection.connect()
            if (connection.responseCode !in 200..399) return null
            val raw = connection.inputStream.use { it.readBytes() }
            val bitmap = BitmapFactory.decodeByteArray(raw, 0, raw.size) ?: return null
            val resized = resizeBitmap(bitmap)
            val output = ByteArrayOutputStream()
            resized.compress(android.graphics.Bitmap.CompressFormat.JPEG, 78, output)
            output.toByteArray() to "image/jpeg"
        } catch (_: Throwable) {
            null
        } finally {
            connection.disconnect()
        }
    }

    private fun resizeBitmap(bitmap: android.graphics.Bitmap): android.graphics.Bitmap {
        val maxDimension = 720f
        val largest = maxOf(bitmap.width, bitmap.height).toFloat()
        if (largest <= maxDimension) return bitmap
        val scale = maxDimension / largest
        return android.graphics.Bitmap.createScaledBitmap(
            bitmap,
            (bitmap.width * scale).toInt(),
            (bitmap.height * scale).toInt(),
            true
        )
    }

    private data class ParsedMetadata(
        val title: String?,
        val description: String?,
        val siteName: String?,
        val canonicalUrl: String?,
        val imageUrl: String?
    )

    private fun parseMetadata(html: String, sourceUrl: String): ParsedMetadata {
        val ogTitle = metaContent(html, property = "og:title")
        val ogDescription = metaContent(html, property = "og:description")
        val ogSiteName = metaContent(html, property = "og:site_name")
        val ogImage = metaContent(html, property = "og:image")
        return ParsedMetadata(
            title = ogTitle ?: titleTag(html),
            description = ogDescription ?: metaContent(html, name = "description"),
            siteName = ogSiteName,
            canonicalUrl = canonicalUrl(html, sourceUrl),
            imageUrl = resolveUrl(ogImage, sourceUrl)
        )
    }

    private fun metaContent(html: String, property: String? = null, name: String? = null): String? {
        val attributeName = if (property != null) "property" else "name"
        val attributeValue = Regex.escape(property ?: name ?: "")
        val patterns = listOf(
            """<meta[^>]*$attributeName\s*=\s*["']$attributeValue["'][^>]*content\s*=\s*["']([^"']+)["'][^>]*>""",
            """<meta[^>]*content\s*=\s*["']([^"']+)["'][^>]*$attributeName\s*=\s*["']$attributeValue["'][^>]*>"""
        )
        return patterns.firstNotNullOfOrNull { pattern ->
            Regex(pattern, setOf(RegexOption.IGNORE_CASE)).find(html)?.groupValues?.getOrNull(1)
        }
    }

    private fun titleTag(html: String): String? {
        return Regex("<title[^>]*>(.*?)</title>", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))
            .find(html)
            ?.groupValues
            ?.getOrNull(1)
    }

    private fun canonicalUrl(html: String, sourceUrl: String): String? {
        val patterns = listOf(
            """<link[^>]*rel\s*=\s*["']canonical["'][^>]*href\s*=\s*["']([^"']+)["'][^>]*>""",
            """<link[^>]*href\s*=\s*["']([^"']+)["'][^>]*rel\s*=\s*["']canonical["'][^>]*>"""
        )
        return patterns.firstNotNullOfOrNull { pattern ->
            Regex(pattern, setOf(RegexOption.IGNORE_CASE))
                .find(html)
                ?.groupValues
                ?.getOrNull(1)
                ?.let { resolveUrl(it, sourceUrl) }
        }
    }

    private fun resolveUrl(raw: String?, sourceUrl: String): String? {
        val value = raw?.trim().orEmpty()
        if (value.isEmpty()) return null
        return runCatching {
            val uri = URI(value)
            if (uri.scheme != null) {
                uri.toString()
            } else {
                URI(sourceUrl).resolve(uri).toString()
            }
        }.getOrNull()
    }

    private fun sanitize(text: String?, limit: Int): String? {
        val decoded = decodeEntities(text.orEmpty())
            .replace(Regex("<[^>]+>"), "")
            .replace(Regex("\\s+"), " ")
            .trim()
        if (decoded.isEmpty()) return null
        return decoded.take(limit)
    }

    private fun decodeEntities(text: String): String {
        return text
            .replace("&amp;", "&")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            .replace("&apos;", "'")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&nbsp;", " ")
    }

    private fun domain(urlString: String): String {
        return runCatching { URI(urlString).host.orEmpty().removePrefix("www.") }.getOrDefault(urlString)
    }
}

File: iOS/Spaces V1.5/Services/LinkPreviewService.swift
import Foundation
import UIKit

struct EncryptedTextMessageContent: Codable {
    let version: Int
    let text: String
    let linkPreview: LinkPreviewData?
    let spaceLinks: [SpaceLinkAttachment]?
}

actor LinkPreviewService {
    static let shared = LinkPreviewService()

    private struct CachedPreviewRecord: Codable {
        let preview: LinkPreviewData
        let cachedAt: Date

        func isExpired(ttl: TimeInterval) -> Bool {
            Date().timeIntervalSince(cachedAt) > ttl
        }
    }

    private let defaults: UserDefaults
    private let session: URLSession
    private var memoryCache: [String: CachedPreviewRecord] = [:]
    private let storageKey = "spaces.linkpreviews.cache"
    private let cacheTTL: TimeInterval = 60 * 60 * 24 * 7

    init(defaults: UserDefaults = .standard, session: URLSession = .shared) {
        self.defaults = defaults
        self.session = session
        self.memoryCache = Self.loadCache(from: defaults, key: storageKey)
    }

    nonisolated static func firstURL(in text: String) -> URL? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        if let detector = try? NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue) {
            let nsRange = NSRange(trimmed.startIndex..<trimmed.endIndex, in: trimmed)
            if let match = detector.matches(in: trimmed, options: [], range: nsRange).first,
               let range = Range(match.range, in: trimmed) {
                let raw = String(trimmed[range])
                return normalizedURL(from: raw)
            }
        }

        for token in trimmed.split(whereSeparator: \.isWhitespace) {
            if let url = normalizedURL(from: String(token)) {
                return url
            }
        }
        return nil
    }

    func cachedPreview(for url: URL) -> LinkPreviewData? {
        let key = Self.cacheKey(for: url)
        guard let record = memoryCache[key], !record.isExpired(ttl: cacheTTL) else {
            memoryCache.removeValue(forKey: key)
            persistCache()
            return nil
        }
        return record.preview
    }

    func preview(for url: URL) async -> LinkPreviewData? {
        let key = Self.cacheKey(for: url)
        if let cached = cachedPreview(for: url) {
            return cached
        }

        do {
            var request = URLRequest(url: url)
            request.timeoutInterval = 15
            request.setValue("Spaces/1.1", forHTTPHeaderField: "User-Agent")
            let (data, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse,
                  (200..<400).contains(httpResponse.statusCode),
                  let html = String(data: data.prefix(256_000), encoding: .utf8) else {
                return nil
            }

            let metadata = Self.parseMetadata(from: html, sourceURL: url)
            let trimmedTitle = metadata.title?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !trimmedTitle.isEmpty else {
                return nil
            }

            let imagePayload = try await loadPreviewImage(
                from: metadata.imageURL,
                sourceURL: url
            )

            let preview = LinkPreviewData(
                originalURL: url.absoluteString,
                canonicalURL: metadata.canonicalURL?.absoluteString,
                domain: Self.domain(from: metadata.canonicalURL ?? url),
                title: Self.sanitizedText(trimmedTitle, limit: 120) ?? trimmedTitle,
                summary: Self.sanitizedText(metadata.description, limit: 220),
                siteName: Self.sanitizedText(metadata.siteName, limit: 80),
                imageDataBase64: imagePayload?.data.base64EncodedString(),
                imageMimeType: imagePayload?.mimeType
            )

            memoryCache[key] = CachedPreviewRecord(preview: preview, cachedAt: Date())
            persistCache()
            return preview
        } catch is CancellationError {
            return nil
        } catch {
            return nil
        }
    }

    private func loadPreviewImage(from candidateURL: URL?, sourceURL: URL) async throws -> (data: Data, mimeType: String)? {
        guard let candidateURL else { return nil }
        let resolvedURL = candidateURL.absoluteURL
        var request = URLRequest(url: resolvedURL)
        request.timeoutInterval = 15
        request.setValue("Spaces/1.1", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200..<400).contains(httpResponse.statusCode),
              data.count > 0 else {
            return nil
        }

        guard let image = UIImage(data: data) else {
            return nil
        }

        let resized = Self.resizedImage(image, maxDimension: 720)
        if let jpeg = resized.jpegData(compressionQuality: 0.78) {
            return (jpeg, "image/jpeg")
        }
        return nil
    }

    private static func loadCache(from defaults: UserDefaults, key: String) -> [String: CachedPreviewRecord] {
        guard let data = defaults.data(forKey: key),
              let decoded = try? JSONDecoder().decode([String: CachedPreviewRecord].self, from: data) else {
            return [:]
        }
        return decoded
    }

    private func persistCache() {
        let filtered = memoryCache.filter { !$0.value.isExpired(ttl: cacheTTL) }
        memoryCache = filtered
        guard let data = try? JSONEncoder().encode(filtered) else { return }
        defaults.set(data, forKey: storageKey)
    }

    private static func normalizedURL(from raw: String) -> URL? {
        let trimmed = raw
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: ".,;:!?)]}\"'"))
        guard !trimmed.isEmpty else { return nil }
        if let direct = URL(string: trimmed), let scheme = direct.scheme?.lowercased(), ["http", "https"].contains(scheme) {
            return direct
        }
        if trimmed.lowercased().hasPrefix("www.") {
            return URL(string: "https://\(trimmed)")
        }
        return nil
    }

    private static func cacheKey(for url: URL) -> String {
        let normalized = (url.scheme?.lowercased() == nil ? "https://" : "") + url.absoluteString.lowercased()
        return normalized
    }

    private static func domain(from url: URL) -> String {
        url.host?.replacingOccurrences(of: #"^www\."#, with: "", options: .regularExpression) ?? url.absoluteString
    }

    private static func resizedImage(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let largest = max(image.size.width, image.size.height)
        guard largest > maxDimension else { return image }
        let scale = maxDimension / largest
        let targetSize = CGSize(width: image.size.width * scale, height: image.size.height * scale)
        let renderer = UIGraphicsImageRenderer(size: targetSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: targetSize))
        }
    }

    private static func sanitizedText(_ text: String?, limit: Int) -> String? {
        guard let text else { return nil }
        let noTags = text.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
        let decoded = decodeHTMLEntities(in: noTags)
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !decoded.isEmpty else { return nil }
        return String(decoded.prefix(limit))
    }

    private static func decodeHTMLEntities(in text: String) -> String {
        text
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: "&apos;", with: "'")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&nbsp;", with: " ")
    }

    private struct ParsedMetadata {
        let title: String?
        let description: String?
        let siteName: String?
        let canonicalURL: URL?
        let imageURL: URL?
    }

    private static func parseMetadata(from html: String, sourceURL: URL) -> ParsedMetadata {
        let ogTitle = metaContent(in: html, property: "og:title")
        let ogDescription = metaContent(in: html, property: "og:description")
        let ogSiteName = metaContent(in: html, property: "og:site_name")
        let ogImage = metaContent(in: html, property: "og:image")
        let title = ogTitle ?? titleTag(in: html)
        let description = ogDescription ?? metaContent(in: html, name: "description")
        let canonical = canonicalURL(in: html, sourceURL: sourceURL)
        let imageURL = resolvedURL(from: ogImage, sourceURL: sourceURL)

        return ParsedMetadata(
            title: title,
            description: description,
            siteName: ogSiteName,
            canonicalURL: canonical,
            imageURL: imageURL
        )
    }

    private static func metaContent(in html: String, property: String? = nil, name: String? = nil) -> String? {
        let attributeName = property != nil ? "property" : "name"
        let attributeValue = NSRegularExpression.escapedPattern(for: property ?? name ?? "")
        let patterns = [
            #"<meta[^>]*\#(attributeName)\s*=\s*["']\#(attributeValue)["'][^>]*content\s*=\s*["']([^"']+)["'][^>]*>"#,
            #"<meta[^>]*content\s*=\s*["']([^"']+)["'][^>]*\#(attributeName)\s*=\s*["']\#(attributeValue)["'][^>]*>"#
        ]

        for pattern in patterns {
            if let match = firstCapture(in: html, pattern: pattern) {
                return match
            }
        }
        return nil
    }

    private static func titleTag(in html: String) -> String? {
        firstCapture(in: html, pattern: #"<title[^>]*>(.*?)</title>"#, options: [.caseInsensitive, .dotMatchesLineSeparators])
    }

    private static func canonicalURL(in html: String, sourceURL: URL) -> URL? {
        let patterns = [
            #"<link[^>]*rel\s*=\s*["']canonical["'][^>]*href\s*=\s*["']([^"']+)["'][^>]*>"#,
            #"<link[^>]*href\s*=\s*["']([^"']+)["'][^>]*rel\s*=\s*["']canonical["'][^>]*>"#
        ]
        for pattern in patterns {
            if let raw = firstCapture(in: html, pattern: pattern),
               let resolved = resolvedURL(from: raw, sourceURL: sourceURL) {
                return resolved
            }
        }
        return nil
    }

    private static func resolvedURL(from raw: String?, sourceURL: URL) -> URL? {
        guard let raw = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else { return nil }
        if let absolute = URL(string: raw), absolute.scheme != nil {
            return absolute
        }
        return URL(string: raw, relativeTo: sourceURL)?.absoluteURL
    }

    private static func firstCapture(
        in text: String,
        pattern: String,
        options: NSRegularExpression.Options = [.caseInsensitive]
    ) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: options) else { return nil }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        guard let match = regex.firstMatch(in: text, options: [], range: range),
              match.numberOfRanges > 1,
              let captureRange = Range(match.range(at: 1), in: text) else {
            return nil
        }
        return String(text[captureRange])
    }
}
